### DISCUSSION TITLE
Old e-mail about ECA-rules

### DISCUSSION THREAD

**Discussion by stefjoosten** [AmpersandTarski/Ampersand#1242] [Ideas]
Sebastiaan wrote the following e-mail, some time ago, which I have not answered yet. This issue is made for not forgetting to do that. It is relevant for the ECA-rule theme in Ampersand. @sjcjoosten : Sebastiaan, if you feel the need to rephrase this question with contemporary insights, please do so in English for @WolframKahl . You can do so by simply editing this comment.

---

Hey Stef,

als je ergens een uurtje de tijd hebt, kun je dan naar onderstaande email kijken?

ik heb eens naar je ECA regels zitten kijken, en heb wat vragen over de structuur:
- een ECArule bestaat uit o.a. een Trigger (Event) en een Delta (Declaration). Een Event bevat weer een Declaration. Moeten deze twee Declarations altijd dezelfde zijn? Zo niet, wat is het verschil?
- een ECArule heeft een getal: ecaNum. Moeten - bij vergelijkingen in Haskell - alle overige velden gelijk zijn als de ecaNum dit is?
- wat voor iets is GCH? Is dit semantisch hetzelfde als een CHC met daarin een lijstje Do's?
- is een ALL met een lege paCls hetzelfde als een Nop? Zo ja, wat is dan hetzelfde als een Blk? GCH of CHC?
- Een ON INSERT d DO INSERT d, is dat óók hetzelfde als een Nop? Wordt hier ook altijd een Nop van gemaakt? En voor ON INSERT d DO DELETE d hetzelfde met Blk?
- wat is de gedachte achter Let ? Wordt dit gebruikt?
- wat is de gedachte achter Ref ? Wordt dit gebruikt?

Verder de observatie dat ik er altijd voor kan zorgen dat alle ECArules voldoende zijn, en er verder geen checks nodig zijn. En wel als volgt:
In regel x komen de declaraties voor. Voor elk van die declaraties d genereer ik deze ECArules:
ON INSERT d BLOCK (motiv: x), ON DELETE d BLOCK (motiv: x). (Of een ALL met daarin een BLOCK) Uiteraard kan je ook slimmere ECArules genereren, dat doe je ongetwijfeld ook, maar ik vroeg me af wat daarbij gebeurt met de eigenschap dat ik alleen maar naar de ECArules hoef te kijken om alle invarianten te behouden. Blijft die eigenschap gelden?

Aan het andere uiterste kan ik ervoor zorgen dat er nooit een actie gebeurt als die niet strikt noodzakelijk is voor het herstellen van een invariant:
ON INSERT d NO-OP (motiv: x), ON DELETE d NO-OP (motiv: x). (Of een CHC/GCH met daarin een NO-OP) Hier geldt dezelfde vraag: als ik alle ECArules toepas zoals hierboven, dan krijg ik geen onnodige INSERTs of DELETEs voor mijn kiezen. Geldt dit ook voor de uiteindelijke set ECArules?

Het lijkt me dat bovenstaande twee eigenschappen niet zonder meer beiden kunnen gelden: ofwel je doet standaard genoeg maar misschien te veel, ofwel je doet niets dat overbodig is maar misschien te weinig. Klopt dit, of is de GCH ervoor om dit probleem op te vangen? En welke van de twee heeft jouw voorkeur / is het belangrijkst?

Groetjes,
Bas

**Comment by sjcjoosten** [comment]
Assigning Stef, so he won't forget to do what he wanted to do (see first Paragraph)
